plugins/catalog-backend-module-ldap/src/ldap/client.ts

```diff
@@ -120,6 +124,53 @@ export class LdapClient {
     }
   }
 
+  /**
+   * Performs an LDAP search operation, calls a function on each entry to limit memory usage
+   *
+   * @param dn The fully qualified base DN to search within
+   * @param options The search options
+   * @param f The callback to call on each search entry
+   */
+  async searchStreaming(
+    dn: string,
+    options: SearchOptions,
+    f: SearchCallback,
+  ): Promise<void> {
+    try {
+      return await new Promise<void>((resolve, reject) => {
+        this.client.search(dn, options, (err, res) => {
+          if (err) {
+            reject(new Error(errorString(err)));
+          }
+
+          res.on('searchReference', () => {
+            reject(new Error('Unable to handle referral'));
+          });
+
+          res.on('searchEntry', async entry => {
+            await f(entry);
```

Comment:
This actually made me stop and think. I wondered how ldapjs actually would handle backpressure. Seems from  a cursory glance that it doesn't.

What I mean by this is, our async callback actually returns (a promise) immediately even though it performs long running work after that in the background. The ldapjs library doesn't await that, so it will pile up work as fast as it can. Our callbacks will start running concurrently with each other and many entries will potentially be live in memory at the same time, and the search call will finish (the returned promise from `searchStreaming` resolves) while callbacks are still doing their thing.

Interesting.

I think that the only safe thing we can do is actually to NOT make the callback async.

Reply:
It's likely due to this being one of my first soiree's into a typescript project, but I was consistently running into an issue where I couldn't properly validate that `entry` existed at execution time. I'll remove this and run some tests to see how it plays out locally

Reply:
How do you feel about the adjusted code? If I understand correctly it should be synchronously handling the callback at this point.